Content.Server/Pinpointer/NavMapSystem.cs

```diff
@@ -384,4 +408,94 @@ public void ToggleBeacon(EntityUid uid, NavMapBeaconComponent? comp = null)
 
         SetBeaconEnabled(uid, !comp.Enabled, comp);
     }
+
+    /// <summary>
+    /// For a given position, tries to find the nearest configurable beacon that is marked as visible.
+    /// This is used for things like announcements where you want to find the closest "landmark" to something.
+    /// </summary>
+    [PublicAPI]
+    public bool TryGetNearestBeacon(Entity<TransformComponent?> ent, [NotNullWhen(true)] out Entity<NavMapBeaconComponent>? beacon)
+    {
+        beacon = null;
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        return TryGetNearestBeacon(_transform.GetMapCoordinates(ent, ent.Comp), out beacon);
+    }
+
+    /// <summary>
+    /// For a given position, tries to find the nearest configurable beacon that is marked as visible.
+    /// This is used for things like announcements where you want to find the closest "landmark" to something.
+    /// </summary>
+    public bool TryGetNearestBeacon(MapCoordinates coordinates, [NotNullWhen(true)] out Entity<NavMapBeaconComponent>? beacon)
+    {
+        beacon = null;
+        var minDistance = float.PositiveInfinity;
+
+        var query = EntityQueryEnumerator<ConfigurableNavMapBeaconComponent, NavMapBeaconComponent, TransformComponent>();
+        while (query.MoveNext(out var uid, out _, out var navBeacon, out var xform))
+        {
+            if (!navBeacon.Enabled)
+                continue;
+
+            if (navBeacon.Text == null)
+                continue;
+
+            if (coordinates.MapId != xform.MapID)
+                continue;
+
+            var coords = _transform.GetWorldPosition(xform);
+            var distanceSquared = (coordinates.Position - coords).LengthSquared();
+            if (!float.IsInfinity(minDistance) && distanceSquared >= minDistance)
+                continue;
+
+            minDistance = distanceSquared;
+            beacon = (uid, navBeacon);
+        }
+
+        return beacon != null;
+    }
+
+    [PublicAPI]
+    public string GetNearestBeaconString(Entity<TransformComponent?> ent)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return Loc.GetString("nav-beacon-pos-no-beacons");
+
+        return GetNearestBeaconString(_transform.GetMapCoordinates(ent, ent.Comp));
+    }
+
+    public string GetNearestBeaconString(MapCoordinates coordinates)
+    {
+        if (!TryGetNearestBeacon(coordinates, out var beacon))
+            return Loc.GetString("nav-beacon-pos-no-beacons");
+
+        var pos = _transform.GetMapCoordinates(beacon.Value);
+
+        var rotOffset = Angle.Zero;
+        if (_mapManager.TryFindGridAt(pos, out var grid, out _))
+            rotOffset = Transform(grid).LocalRotation;
+
+        var dir = (pos.Position - coordinates.Position).GetDir();
+        var adjustedDir = (dir.ToAngle() - rotOffset + (Math.PI)).GetDir();
+
```

Comment:
Is this supposed to be worldangle, you didn't even document what this is trying to do.